pwndbg/aglib/vmmap_custom.py

```diff
@@ -0,0 +1,154 @@
+    Returns a tuple of `Page` objects representing the memory mappings of the
+    target, sorted by virtual address ascending.
+    """
+    pages: List[pwndbg.lib.memory.Page] = [*explored_pages, *custom_pages]
+    pages.sort()
+    return tuple(pages)
+
+
+def add_custom_page(page: pwndbg.lib.memory.Page) -> None:
+    bisect.insort(custom_pages, page)
+
+    # Reset all the cache
+    # We can not reset get() only, since the result may be used by others.
+    # TODO: avoid flush all caches
+    pwndbg.lib.cache.clear_caches()
+
+
+def clear_custom_page() -> None:
+    while custom_pages:
+        custom_pages.pop()
+
+    # Reset all the cache
+    # We can not reset get() only, since the result may be used by others.
+    # TODO: avoid flush all caches
+    pwndbg.lib.cache.clear_caches()
+
+
+def explore(address_maybe: int) -> pwndbg.lib.memory.Page | None:
+    """
+    Given a potential address, check to see what permissions it has.
+
+    Returns:
+        Page object
+
+    Note:
+        Adds the Page object to a persistent list of pages which are
+        only reset when the process dies. This means pages which are
+        added this way will not be removed when unmapped.
+
+        Also assumes the entire contiguous section has the same permission.
+    """
+    if not pwndbg.dbg.selected_inferior().is_linux():
+        return None
+
+    if auto_explore.value == "warn":
+        page_start = pwndbg.lib.memory.page_align(address_maybe)
+        if page_start not in _warn_cache:
+            _warn_cache.add(page_start)
+            is_readable_addr = pwndbg.aglib.memory.peek(page_start)
+            if is_readable_addr:
+                print(
+                    M.warn(
+                        f"Warning: Avoided exploring possible address {address_maybe:#x}.\n"
+                        f"You can explicitly explore it with `vmmap_explore {page_start:#x}`"
+                    )
+                )
+        return None
+    elif auto_explore.value == "no":
+        return None
+
+    address_maybe = pwndbg.lib.memory.page_align(address_maybe)
+    flags = get_memory_flags(address_maybe)
+    if flags is None:
+        return None
+
+    page = find_boundaries(address_maybe)
+    page.objfile = "<explored>"
+    page.flags = flags
+
+    explored_pages.append(page)
+
+    # Reset all the cache
+    # We can not reset get() only, since the result may be used by others.
+    # TODO: avoid flush all caches
+    pwndbg.lib.cache.clear_caches()
+
+    return page
+
+
+def get_memory_flags(address_maybe: int) -> int | None:
+    flags = 4 if pwndbg.aglib.memory.peek(address_maybe) else 0
+    if not flags:
+        return None
+
+    if pwndbg.aglib.memory.poke(address_maybe):
+        flags |= 2
+    # It's really hard to check for executability, so we just make some guesses:
+    # If it's in the same page as the stack pointer, try to check the NX bit
+    # If it's in the same page as the instruction pointer, assume it's executable
```

Comment:
Please reorder this line with the above one, since we first check for pc register and then for sp+NX bit

Reply:
done